torch/nn/modules/loss.py

```diff
@@ -1254,6 +1257,108 @@ def forward(self, anchor: Tensor, positive: Tensor, negative: Tensor) -> Tensor:
                                      eps=self.eps, swap=self.swap, reduction=self.reduction)
 
 
+class TripletMarginLossWithDistance(_Loss):
+    r"""Creates a criterion that measures the triplet loss given an input
+    tensors :math:`x1`, :math:`x2`, :math:`x3` and a margin with a value greater than :math:`0`.
+    This is used for measuring a relative similarity between samples. A triplet
+    is composed by `a`, `p` and `n` (i.e., `anchor`, `positive examples` and `negative
+    examples` respectively). The shapes of all input tensors should be
```

Comment:
This shape restriction is really interesting. Why is it required?

Reply:
That does sound better. I think one of the advantages of building your own distance function is you can support a wider variety of shapes. Maybe something more like, "tensors can be any size that the provided distance function can handle?"